go-api/utils/custom_types.go

```diff
@@ -0,0 +1,257 @@
+		*ci_32 = CustomInt32(int32(v))
+	case string:
+		if v == "" {
+			*ci_32 = CustomInt32(0)
+		} else {
+			converted, err := strconv.Atoi(v)
+			if err != nil {
+				return err
+			}
+			*ci_32 = CustomInt32(int32(converted))
+		}
+
+	default:
+		return fmt.Errorf("unexpected type for customInt32: %T", value)
+	}
+	return nil
+}
+
+func (ci_64 CustomInt64) String() string {
+	return strconv.FormatInt(int64(ci_64), 10)
+}
+
+func (ci_64 *CustomInt64) MarshalJSON() ([]byte, error) {
+
+	return json.Marshal(ci_64.String())
+}
+
+func (ci_64 *CustomInt64) UnmarshalJSON(data []byte) error {
+	var value interface{}
+	if err := json.Unmarshal(data, &value); err != nil {
+		return err
+	}
+
+	switch v := value.(type) {
+	case int64:
+		*ci_64 = CustomInt64(v)
+	case int32:
+		*ci_64 = CustomInt64(int64(v))
+	case int:
+		*ci_64 = CustomInt64(int64(v))
+	case float64:
+		*ci_64 = CustomInt64(int64(v))
+	case float32:
+		*ci_64 = CustomInt64(int64(v))
+	case string:
+		if v == "" {
+			*ci_64 = CustomInt64(0)
+		} else {
+			converted, err := strconv.Atoi(v)
+			if err != nil {
+				return err
+			}
+			*ci_64 = CustomInt64(converted)
+		}
+
+	default:
+		return fmt.Errorf("unexpected type for CustomInt64: %T", value)
+	}
+	return nil
+}
+
+func (cdt CustomDateTime) String() string {
+	utcTime := cdt.Time.UTC()
+	return utcTime.Format(RFC3339Milli)
+}
+
+func (cdt *CustomDateTime) MarshalJSON() ([]byte, error) {
+	return json.Marshal(cdt.String())
+}
+
+func (cdt *CustomDateTime) Scan(src interface{}) error {
+	switch v := src.(type) {
+	case time.Time:
+		cdt.Time = v
+	case string:
+		// https://github.com/golang/go/issues/40472
+		// Preprocess the time zone format
+		v = strings.Replace(v, "GMT", "UTC", -1)
+
+		// Attempt to parse the string using the specified formats
+		layouts := []string{"Mon Jan 02 2006 15:04:05 MST-0700", "Jan 02 2006 15:04:05 MST-0700", "Mon Jan 02 2006 15:04:05 MST-07", "Jan 02 2006 15:04:05 MST-07"}
+		var converted time.Time
+		var err error
+
+		for _, layout := range layouts {
+			converted, err = time.Parse(layout, v)
+			if err == nil {
+				break
+			}
+		}
+
+		// If all formats fail, try using dateparse
+		if err != nil {
+			converted, err = dateparse.ParseAny(v)
+			if err != nil {
+				log.Panicf("Failed to parse:", v)
+				return err
+			}
+		}
+
+		cdt.Time = converted
+	default:
+		return fmt.Errorf("unexpected type for CustomDateTime: %T", src)
+	}
+	return nil
+}
+
+func (cdt *CustomDateTime) UnmarshalJSON(data []byte) error {
+	var value interface{}
+	if err := json.Unmarshal(data, &value); err != nil {
+		return err
+	}
+
+	switch v := value.(type) {
+	case time.Time:
+		cdt.Time = v
+	case string:
+		// Preprocess the time zone format
+		v = strings.Replace(v, "GMT", "UTC", -1)
+
+		// Attempt to parse the string using the specified formats
+		layouts := []string{"Mon Jan 02 2006 15:04:05 MST-0700", "Jan 02 2006 15:04:05 MST-0700", "Mon Jan 02 2006 15:04:05 MST-07", "Jan 02 2006 15:04:05 MST-07"}
```

Comment:
@nick-bisonai Did you encounter date formats that differ from each other in this API? 🤔

Reply:
I'm assuming that most of inputs currently follows next format
`Jan 02 2006 15:04:05 GMT-0700`

but tried to catch more cases just in case. If only certain case are handled the code could be simplified, but I tried rather safer way

Reply:
I understand what you mean, but I have opposite opinion 😊 We don't want to happen that we use multiple different formats to describe the same underlying data type. Having multiple options shows that we are open to accepting any format, but it is actually not true, we want just want one. If we keep it we might end up like Bithumb's websocket API 🤦

Reply:
what kind of problem does Bithumb's websocket API have? is it bad to take multiple formats even though it's saved in single type?

Reply:
In Bithumb's API they return date in at least three different formats.

ticker
```
{
    "date" : "20200129",
    "time" : "121844"
}
```

transaction
```
{
    "contDtm" : "2020-01-29 12:24:18.830039"
}
```

orderbooksnapshot

```
{
     "datetime": "1680082623245840"
}
```

It is not necessarily bad if we accept date/time in various formats, but the point is we don't have to be nice, we have to be strict and straightforward. If there is API which allows multiple different data types and the application takes advantage of it, it might lead to multiple different implementations for date/time format. New developers will be confused, people might opt for reimplementation of date formatting and it will be hard to tell them to not to do it, because the API allows it. Usually conversion from date/time to string has also implementation of conversion in the opposite direction, and then if some parts of API will be implemented by different people, they will prefer to return the time in their favorite format (as in Bithumb's case).

Reply:
so `UnmarshalJSON` is used for just reading json formatted dateTime type, and in our case even though it allows multiple dateTime types to read, it only has one type on returning dateTime format

```
func (cdt CustomDateTime) String() string {
	utcTime := cdt.Time.UTC()
	return utcTime.Format(RFC3339Milli)
}

func (cdt *CustomDateTime) MarshalJSON() ([]byte, error) {
	return json.Marshal(cdt.String())
}
```

whenever it returns data it'll only return in single format

Reply:
awh so what you're saying is it should be intentionally restricted to how it's returned
i see

Reply:
Yes. Additionally, having less code is always better. We don't have to prepare for situations that did not come up or we don't see why they would come up :)

Reply:
nestJs didn't have that kind of strict design

when inputting dateTime, string has following format
`Jan 02 2006 15:04:05 GMT-0700`

when returning dateTIme, string has following format 
`2006-01-02T15:04:05.000Z07:00`

how should I set things up from go api?

Reply:
Very interesting! I think we should be strict. Let's choose one of them. I am not sure about the Data Feed worker (it uses time for checking data staleness), but I think the second is better machine readable 🤔

Reply:
The first dateTime string type is generated from other services rather than api, so if choosing one there should be an updates in other services as well

Reply:
Aha! So then please just identify what service we receive dates from and let's make the correction in subsequent PR.